<h1 id="objectives">Objectives</h1>
<p>The learning objectives for this practical are:</p>
<ul>
<li>Reading DNA from FASTA files.</li>
<li>Implement a program in Python that calculates the percentage of CpG dinucleotides in a DNA sequence stored in a FASTA file.</li>
<li>Implement a program in Python that calculates the number of <code>CAG</code> tri-nucleotides, and the position of the first <code>CAG</code> tri-nucleotide found, in a DNA sequence stored in a FASTA file.</li>
<li>Make syntax errors in Python.</li>
<li>Correct syntax errors in Python.</li>
<li>Debug your program when it doesn’t work.</li>
</ul>
<p>Whenever you are stuck with an error, please consult the section entitled “Debugging” from <a href="/practical4#debugging">practical 4</a>.</p>
<h1 id="setup-and-background">Setup and background</h1>
<p>To do this practical you need an installation of Python version 3. You can find the instructions in the <a href="/setup/">setup</a> link on how to install Python version 3 in your system. Once Python is installed, you should be able to call it from the shell in the terminal window. You can check whether that is possible by typing:</p>
<pre><code>$ which python
$ python --version</code></pre>
<p>It may happen that you have two Python installations, one corresponding to version 2.x and another to version 3.x. In that situation the previous command may say that your Python version is 2.x and to access the version 3 you need to call the executable <code>python3</code>. Try then for instance:</p>
<pre><code>$ python3 --version</code></pre>
<p>If this is your case, then whenever the executable <code>python</code> is invoked in the rest of this practical, please use <code>python3</code> instead.</p>
<h1 id="reading-dna-from-fasta-files">Reading DNA from FASTA files</h1>
<p>Following the explanation from Wikipedia, in bioinformatics the <a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA format</a> is a text-based file format for representing sequences of either <a href="https://en.wikipedia.org/wiki/Nucleic_acid_sequence">nucleic acids</a> or <a href="https://en.wikipedia.org/wiki/Amino_acid">amino acids</a>. Here, we want to learn how can we read from a Python program the DNA sequence of a gene stored in FASTA format within a text file and make some simple summaries of the DNA content from that gene. We are going to illustrate that task with the <a href="https://en.wikipedia.org/wiki/Hemoglobin_subunit_beta"><em>hemoglobin subunit beta</em> (HBB) gene</a>, which is a gene coding for a protein that forms part of <a href="https://en.wikipedia.org/wiki/Hemoglobin">haemoglobin</a>, the molecule responsible for transporting oxygen in <a href="https://en.wikipedia.org/wiki/Red_blood_cell">red blood cells</a> of almost all <a href="https://en.wikipedia.org/wiki/Vertebrate">vertebrates</a>.</p>
<p>First, let’s download the FASTA file of the DNA of the human <em>HBB</em> gene following these steps:</p>
<ol type="1">
<li>Go to the NCBI web page for the human <em>HBB</em> gene at this <a href="https://www.ncbi.nlm.nih.gov/gene/3043">link</a>.</li>
<li><p>Scroll down to the section entitled “Genomic regions, transcripts and products” and click on the link called <code>FASTA</code>, as indicated by the red arrow in the following image.</p>
<p><img src="ncbigene1.jpg" /></p></li>
<li><p>In the next page, click on the link called <code>Send to:</code>, from the pull-down menu click on <code>File</code> and finally on the button <code>Create File</code>, as indicated by the red arrows in the following image.</p>
<p><img src="ncbigene2.jpg" /></p></li>
<li>A file called <code>sequence.fasta</code> will be downloaded from your browser, probably into your downloads directory. Move it into the folder that you are using for this practical under the name <code>HBB.fa</code>. This is the FASTA file containing the DNA of the human <em>HBB</em> gene.</li>
<li><p>Examine the first lines of this FASTA file with the Unix <code>head</code> command. The result should be as follows:</p>
<pre><code>$ head HBB.fa
&gt;NC_000011.10:c5227071-5225464 Homo sapiens chromosome 11, GRCh38.p14 Primary Assembly
ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGGTGCATCTGACTCCTGA
GGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGC
AGGTTGGTATCAAGGTTACAAGACAGGTTTAAGGAGACCAATAGAAACTGGGCATGTGGAGACAGAGAAG
ACTCTTGGGTTTCTGATAGGCACTGACTCTCTCTGCCTATTGGTCTATTTTCCCACCCTTAGGCTGCTGG
TGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGGG
CAACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTCGGTGCCTTTAGTGATGGCCTGGCTCACCTGGAC
AACCTCAAGGGCACCTTTGCCACACTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGATCCTGAGAACT
TCAGGGTGAGTCTATGGGACGCTTGATGTTTTCTTTCCCCTTCTTTTCTATGGTTAAGTTCATGTCATAG
GAAGGGGATAAGTAACAGGGTACAGTTTAGAATGGGAAACAGACGAATGATTGCATCAGTGTGGAAGTCT</code></pre></li>
</ol>
<p>Next, create a new text file called <code>tallynt.py</code> with the following Python program:</p>
<pre><code>f = open(&quot;HBB.fa&quot;)
line = f.readline()                  ## read the first line from HBB.fa
print(&quot;The DNA sequence from gene:&quot;)
print(line.strip())                  ## print the first line from HBB.fa

seq = &quot;&quot;                             ## seq will store the whole gene DNA
while (line) :                       ## while &#39;line&#39; is not empty
    line = f.readline()              ## read the next line from HBB.fa
    seq = seq + line.strip()         ## concatenate that line to &#39;seq&#39;

v = list(seq)                        ## convert &#39;seq&#39; into a vector &#39;v&#39;
n = len(v)                           ## calculate the length of vector &#39;v&#39;
print(f&quot;has a total of {n} nucleotides&quot;)</code></pre>
<p>This Python program reads the lines from <code>HBB.fa</code> and concanates them into a single character string called <code>seq</code>, which afterwards is converted into a vector <code>v</code>, from which we calculate its length, corresponding to the total number of nucleotides in the DNA encoding the gene <em>HBB</em>. It includes a call to the following functions:</p>
<ul>
<li><a href="https://www.w3schools.com/python/ref_func_open.asp"><code>open()</code></a>, which opens a text file and returns an object that allows one to read lines from that file using the method <a href="https://www.w3schools.com/python/ref_file_readline.asp"><code>f.readline()</code></a>, for some file object <code>f</code>.</li>
<li><a href="https://www.w3schools.com/python/ref_string_strip.asp"><code>str.strip()</code></a>, where <code>str</code> is some character string object and it removes any leading and trailing spaces and newlines from that string.</li>
</ul>
<p>When you run this program, the output must be as follows:</p>
<pre><code>$ python tallynt.py
The DNA sequence from gene:
&gt;NC_000011.10:c5227071-5225464 Homo sapiens chromosome 11, GRCh38.p14 Primary Assembly
has a total of 1608 nucleotides</code></pre>
<p>Notice that you could also count the nucleotides using a Unix shell command line as follows:</p>
<pre><code>$ grep -v &#39;&gt;&#39; HBB.fa | fold -1 | wc -l
1609</code></pre>
<p>Try to understand what is doing each bit of the previous Unix command-line. Could you figure out why this command line is giving a number that is one unit larger than what our Python programe gives?</p>
<ol type="1">
<li><p>Enable the program to take the name of the FASTA file as a first argument in the command line, so that it can work with any FASTA file containing DNA from any gene. Run it, for instance, with the DNA from the the <a href="https://www.ncbi.nlm.nih.gov/gene/59272">angiotensin converting enzyme 2 (<em>ACE2</em>) gene</a>, which encodes a protein that acts as a <a href="https://www.nytimes.com/interactive/2020/03/11/science/how-coronavirus-hijacks-your-cells.html">receptor for the spike glycoprotein of the human coronavirus SARS-CoV-2</a>, the causative agent of coronavirus disease-2019 (COVID-19), e.g.:</p>
<pre><code>$ python tallynt.py ACE2.fa</code></pre></li>
<li><p>Enable the program to take a second argument where we specify one of the four possible nucleotides (<code>A</code>, <code>C</code>, <code>G</code> or <code>T</code>) and it calculates the number of occurrences of that nucleotide in the DNA sequence of the gene, e.g.:</p>
<pre><code>$ python tallynt.py HBB.fa A</code></pre>
<p>You can verify whether the calculation is correct by doing it also in Unix with the following command line:</p>
<pre><code>$ grep -v &#39;&gt;&#39; HBB.fa | fold -1 | grep A | wc -l
411</code></pre>
<p>Try to understand what is doing each bit of the previous Unix shell command line. Could you think of a way to tally all four nucleotides in the same command line?</p></li>
</ol>
<h1 id="cpg-dinucleotides">CpG dinucleotides</h1>
<p>The Wikipedia <a href="https://en.wikipedia.org/wiki/DNA">page</a> for DNA says:</p>
<blockquote>
<p>Deoxyribonucleic acid (DNA) is a polymer composed of two polynucleotide chains that coil around each other to form a double helix.</p>
</blockquote>
<p>The nucleotides forming DNA are adenine (denoted by <code>A</code>), cytosine (<code>C</code>), guanine (<code>G</code>) and thymine (<code>T</code>). The Wikipedia <a href="https://en.wikipedia.org/wiki/CpG_site">page</a> for a CpG site says:</p>
<blockquote>
<p>CpG is shorthand for 5’—C—phosphate—G—3’ , that is, cytosine and guanine separated by only one phosphate group; phosphate links any two nucleosides together in DNA. The CpG notation is used to distinguish this single-stranded linear sequence from the CG base-pairing of cytosine and guanine for double-stranded sequences. […] CpG dinucleotides have long been observed to occur with a much lower frequency in the sequence of vertebrate genomes than would be expected due to random chance. […] This underrepresentation is a consequence of the high mutation rate of methylated CpG sites. […] CpG islands (or CG islands) are regions with a high frequency of CpG sites. […] CpG islands typically occur at or near the transcription start site of genes, particularly housekeeping genes, in vertebrates. […] In cancers, loss of expression of genes occurs about 10 times more frequently by hypermethylation of promoter CpG islands than by mutations.</p>
</blockquote>
<p>Implement a program in Python that given the name of a FASTA file containing the DNA sequence of a gene, it calculates the percentage of <code>CpG</code> sites for every line of DNA in the FASTA file. It is important that the program does <em>not</em> ask for the name of the FASTA file, but instead it takes it as the first argument from the Unix shell command-line call, i.e., by doing something like:</p>
<pre><code>$ python cpg.py &lt;filename.fa&gt;</code></pre>
<p>Try your program with the DNA of the <a href="https://www.ncbi.nlm.nih.gov/gene/672">Breast cancer type 1 susceptibility gene (<em>BRCA1</em>)</a>, a tumor suppressor gene encoding a DNA repair enzyme that becomes hypermetilated in cancer and whose <a href="https://en.wikipedia.org/wiki/BRCA1">role in breast cancer susceptibility</a> was discovered by <a href="https://en.wikipedia.org/wiki/Mary-Claire_King">Mary-Claire King</a> (in your spare time, watch this 12-minute <a href="https://www.youtube.com/watch?v=tOP5pUIYhv4">video</a> where she explains the few days before she was defending in front of a panel at the <a href="https://www.nih.gov">NIH</a> the grant application that was ultimately approved and allowed her to develop her research on the role of the <em>BRCA1</em> gene as a breast cancer susceptibility gene). The first 10 lines of output of your program should be the following percentage values:</p>
<pre><code>$ python cpg.py BRCA.fa
5.7
11.4
14.3
20.0
2.9
2.9
17.1
5.7
11.4
11.4</code></pre>
<p>Note that to get one decimal digits you should use the format specifier <code>{z:.1f}</code>, if <code>z</code> would be the name of the variable with the result, in the formatted string literal that you give to the print function (see the explanation about format strings in section “Adding two numbers” from <a href="https://funcompbio.github.io/practical4/#adding-two-numbers">practical 4</a>).</p>
<p>In <a href="https://funcompbio.github.io/practical5/#reading-dna-from-fasta-files">practical 5</a> you will find the steps on how to download the FASTA file of the DNA of a given gene. Try also to have a modular design of your program, for instance by having a function for the calculation of the percentage of <code>CGs</code> given a vector of nucleotides.</p>
<h1 id="cag-tri-nucleotides">CAG tri-nucleotides</h1>
<p><a href="https://en.wikipedia.org/wiki/Huntington%27s_disease">Huntington’s disease</a> is an hereditary genetic disorder, caused by an expansion of consecutive repetitions of the <code>CAG</code> tri-nucleotide in a gene named after the disease, the <a href="https://www.ncbi.nlm.nih.gov/gene/3064">Huntingtin (<em>HTT</em>)</a> gene. Unaffected individuals usually have no more than 30 consecutive repetitions of the <code>CAG</code> tri-nucleotide in the <em>HTT</em> gene, while affected individuals usually have more than 37.</p>
<p>Implement a program in Python that given the name of a FASTA file containing the DNA sequence of a gene, it calculates the total number of <code>CAG</code> tri-nucleotides and the position of the first <code>CAG</code> tri-nucleotide found in the DNA sequence. It is important that the program does <em>not</em> ask for the name of the FASTA file, but instead it takes it as the first argument from the Unix shell command-line call, i.e., by doing something like:</p>
<pre><code>$ python cags.py &lt;filename&gt;</code></pre>
<p>Try your program with the DNA of the <em>HTT</em> gene. The output for the <em>HTT</em> gene should be this one:</p>
<pre><code>$ python cags.py HTT.fa
3982 CAG tri-nucleotides
first CAG at position 33</code></pre>
<p>Note that this is a simplified version of the biological question. Your program needs not to calculate the number of <em>consecutive</em> <code>CAG</code> tri-nucleotides, but simply the total number of <code>CAG</code> tri-nucleotides.</p>
